tests/MongoDB.Driver.Tests/AggregateGraphLookupTests.cs

```diff
@@ -0,0 +1,247 @@
+                .ToList();
+
+            result.Count.Should().Be(2);
+            result[0].ToBsonDocument().Should().Be(expectedResult[0].ToBsonDocument());
+            result[1].ToBsonDocument().Should().Be(expectedResult[1].ToBsonDocument());
+        }
+
+        // private methods
+        private void EnsureTestDataA(IMongoDatabase database, string collectionName)
+        {
+            database.DropCollection(collectionName);
+            var collection = database.GetCollection<A>(collectionName);
+            var documents = new A[]
+            {
+                new A { From = new X(1), To = new X(2) },
+                new A { From = new X(2), To = new X(3) },
+            };
+            collection.InsertMany(documents);
+        }
+
+        private void EnsureTestDataB(IMongoDatabase database, string collectionName)
+        {
+            database.DropCollection(collectionName);
+            var collection = database.GetCollection<B>(collectionName);
+            var documents = new B[]
+            {
+                new B { From = new X(1), To = new X[] { new X(2), new X(3) } },
+                new B { From = new X(2), To = new X[] { new X(3), new X(4) } },
+            };
+            collection.InsertMany(documents);
+        }
+
+        private void EnsureTestDataC(IMongoDatabase database, string collectionName)
+        {
+            database.DropCollection(collectionName);
+            var collection = database.GetCollection<C>(collectionName);
+            var documents = new C[]
+            {
+                new C { From = new X[] { new X(2), new X(3) }, To = new X(1) },
+                new C { From = new X[] { new X(3), new X(4) }, To = new X(2) },
+            };
+            collection.InsertMany(documents);
+        }
+
+        // nested types
+        [BsonIgnoreExtraElements]
+        private class A
+        {
+            public X From { get; set; }
+            public X To { get; set; }
+        }
+
+        [BsonIgnoreExtraElements]
+        private class AMap
+        {
+            public X From { get; set; }
+            public X To { get; set; }
+            public List<A> Map { get; set; }
+        }
+
+        [BsonIgnoreExtraElements]
+        private class B
+        {
+            public X From { get; set; }
+            public IEnumerable<X> To { get; set; }
+        }
+
+        [BsonIgnoreExtraElements]
+        private class BMap
+        {
+            public X From { get; set; }
+            public IEnumerable<X> To { get; set; }
+            public List<B> Map { get; set; }
+        }
+
+        [BsonIgnoreExtraElements]
+        private class C
+        {
+            public IEnumerable<X> From { get; set; }
+            public X To { get; set; }
+        }
+
+        [BsonIgnoreExtraElements]
+        private class CMap
+        {
+            public IEnumerable<X> From { get; set; }
+            public X To { get; set; }
+            public List<C> Map { get; set; }
+        }
```

Comment:
Replace `CMap` with:

    private class ManyToOneResult
    {
        public int Id { get; set; }
        public IEnumerable<int> From { get; set; }
        public int To { get; set; }
        public List<ManyToOne> Matches { get; set; }
    }